mitodl/open-discussions
# Featured list, lists and subfields for FieldChannel

#### Pre-Flight checklist

#### What are the relevant tickets?
Part of https://github.com/mitodl/open-discussions/issues/3633
Part of #3632 

#### What's this PR do?
- Adds a few new attributes to `FieldChannel` (`lists`, `featured_list`, `subfields`) as well as new models to support ordering of lists and subfields per `FieldChannel`
- Modifies serializers to support adding/modifying these new attributes
- Moves avatar/banner serializer fields into a separate `ChannelAppearanceMixin` used by `ChannelSerializer` and `FieldChannelWriteSerializer`
- Modifies the `UserListViewSet` to optionally show all public lists if a `public=True` request parameter is included (otherwise it only shows lists created by the user).
- Bases common channel fields off of the `discussions.models.Channel` field instead of `channels.models.Channel` (hadn't realized or maybe forgot this existed, see https://github.com/mitodl/open-discussions/pull/2758)

#### How should this be manually tested?
- Go to `http://localhost:8063/api/v0/fields/` while logged in as an admin user, if there are already some existing field channels then you should see them listed.
- Create several new public user lists from `http://localhost:8063/learn` (click the bookmark icon for any ocw course, click 'Create New List', fill out the form using 'Public' for privacy, and save).  Create a private list too.
- Create a new `FieldChannel` by posting the following JSON, you should get a valid response.
    ```
    {
        "name": "new_field_1",
        "title": "New Field 1",
        "description": "New Field 1",
    }
   ```

- Go to `http://localhost:8063/api/v0/fields/new_field_1`, the details on your new field should be listed.
- Create another new `FieldChannel` with more data by posting the following JSON, you should get a valid response.
    ```
    {
        "name": "new_field_2",
        "title": "New Field 2",
        "description": "New Field 2",
        "lists": [<public_list_1_id>, <public_list_2_id>],
        "featured_list": <public_list_1_id>,
        "subfields": ["new_field_1"]
    }
   ```

- Go to `http://localhost:8063/api/v0/fields/new_field_2`, the details on your new field should be listed, including lists and subfields.

- PATCH the following JSON to update the data including lists and subfields:
    ```
    {
        "title": "New Field 2 modified",
        "description": "New Field 2 new desc",
        "lists": [<public_list_2_id>],
        "featured_list": <public_list_2_id>,
        "subfields": []
    }
    ```
- Reload the detail page and verify that the fields have been correctly updated
- PATCH the following JSON to try updating the `lists` field with a private user list, you should get a validation error:
```
    {
        "lists": [<private_list_id>]
    }
```

- Log in as a different user and create some more private and public lists.
- Go to `http://localhost:8063/api/v0/userlists/`, you should see only that user's public and private lists
- Go to `http://localhost:8063/api/v0/userlists/?public=true`, you should see all users' public lists
